### PR TITLE
JDK-8297408: Consolidate code in runtime/ErrorHandling

### PR DESCRIPTION
**Pull request by tstuefe**
There is a lot of duplicate coding in runtime/ErrorHandling that could be cut down using the new HsErrFileUtils utilities added with [JDK-8296906](https://bugs.openjdk.org/browse/JDK-8296906).

This patch mainly cuts away duplicate coding. Very minor functional changes:
- All tests will now use try-with-resources when opening hs-err files
- [TestCrashOnOutOfMemoryError.java](https://github.com/openjdk/jdk/pull/11283/files#diff-7560f2e2700932e015563716b3b0880495922c4c5731ddadba3b33bd937358bd) now checks for at least one matching line in the hs-err file - before, it basically just scanned for the END marker, which seemed pointless
- [test/hotspot/jtreg/runtime/ErrorHandling/TestDwarf.java](https://github.com/openjdk/jdk/pull/11283/files#diff-c04dd18d5a29a42e3a9a008c0b917663a7617c8c9a0d6e9efe8755773260ea95) : heaved Pattern construction out of loop
- [test/hotspot/jtreg/runtime/ErrorHandling/TimeoutInErrorHandlingTest.java](https://github.com/openjdk/jdk/pull/11283/files#diff-086956c42403a16043a111b0582736ac21ce0522dba866f5f9bc3fdb16037a0c) : I tried to preserve the debug output that was painstakingly added by @dcubed-ojdk apart from one change: I now printe the hs-err file as part of the scanning process. This works fine since in case of success, it will only be parsed as far as all patterns are matched, which happens after five-ish lines. Only if none are found, we scan and therefore print the whole hs-err file. So the old behavior is preserved in spirit.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297408](https://bugs.openjdk.org/browse/JDK-8297408): Consolidate code in runtime/ErrorHandling

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [b9be700a](https://git.openjdk.org/jdk/pull/11283/files/b9be700a3f4c00feada2f5c0418cd1f7ca1f7dd5)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**) ⚠️ Review applies to [85996592](https://git.openjdk.org/jdk/pull/11283/files/85996592af82e035d29cfc71c29237caee5cdeac)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11283/head:pull/11283` \
`$ git checkout pull/11283`

Update a local copy of the PR: \
`$ git checkout pull/11283` \
`$ git pull https://git.openjdk.org/jdk pull/11283/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11283`

View PR using the GUI difftool: \
`$ git pr show -t 11283`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11283.diff">https://git.openjdk.org/jdk/pull/11283.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java [hunk 1]
```diff
@@ -60,46 +61,86 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
     }
 
     /**
-     * Given an open hs-err file, read it line by line and check for pattern. Pattern
-     * need to appear in order, but not necessarily uninterrupted.
+     * Given an open hs-err file, read it line by line and check for existence of a pattern. Will fail
+     * if pattern are missing, or if the END marker is missing.
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:65]: "patterns are".

File: test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java [hunk 2]
```diff
@@ -60,46 +61,86 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
     }
 
     /**
-     * Given an open hs-err file, read it line by line and check for pattern. Pattern
-     * need to appear in order, but not necessarily uninterrupted.
+     * Given an open hs-err file, read it line by line and check for existence of a pattern. Will fail
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:64]: "a set of patterns".

File: test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java [hunk 3]
```diff
@@ -60,46 +61,86 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
...
-
-        String lastLine = null;
-        while ((line = br.readLine()) != null) {
-            if (verbose) {
-                System.out.println(line);
+    /**
+     * Given an open hs-err file, read it line by line and check for various conditions.
+     * @param f input file
+     * @param positivePatterns Optional array of patterns that need to appear, in given order, in the file. Missing
+     *                        pattern cause the test to fail.
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:80]: patterns

File: test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java [hunk 4]
```diff
@@ -60,46 +61,86 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
...
+     * @param negativePatterns Optional array of patterns that must not appear in the file; test fails if they do.
+     *                        Order is irrelevant.
+     * @param checkEndMarker If true, we check for the final "END" in an hs-err file; it missing indicates hs-err file
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:83]: "if it is missing it indicates"

File: test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java [hunk 5]
```diff
@@ -60,46 +61,86 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
...
-        int currentPattern = 0;
-
-        String lastLine = null;
-        while ((line = br.readLine()) != null) {
-            if (verbose) {
-                System.out.println(line);
+    /**
+     * Given an open hs-err file, read it line by line and check for various conditions.
+     * @param f input file
+     * @param positivePatterns Optional array of patterns that need to appear, in given order, in the file. Missing
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:79]: Consistency check: Sometimes you start the description with a capital letter and other times not. I don't know what preferred JDK Java style is.

File: test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java [hunk 6]
```diff
@@ -60,46 +61,86 @@ public static File openHsErrFileFromOutput(OutputAnalyzer output) {
...
+                            }
+                            System.out.println("^^^ Forbidden pattern found at line " + lineNo + ": " + negativePattern + "^^^");
+                            throw new RuntimeException("Forbidden pattern found at line " + lineNo + ": " + negativePattern);
+                        }
                     }
-                    System.out.println("^^^ Match " + currentPattern + ": matches " + patterns[currentPattern] + "^^^");
-                    currentPattern ++;
                 }
+                lastLine = line;
+                lineNo ++;
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java:133]: No space before unary operators.

File: test/hotspot/jtreg/runtime/ErrorHandling/TestDwarf.java [hunk 7]
```diff
@@ -123,18 +123,15 @@ private static void runAndCheck(Flags flags, DwarfConstraint... constraints) thr
...
-        Matcher matcher = pattern.matcher(crashOutputString);
         System.out.println(crashOutputString);
-        if (matcher.find()) {
-            String hsErrFileName = matcher.group();
-            System.out.println("hs_err_file: " + hsErrFileName);
-            File hs_err_file = new File(hsErrFileName);
-            BufferedReader reader = new BufferedReader(new FileReader(hs_err_file));
+
+        File hs_err_file = HsErrFileUtils.openHsErrFileFromOutput(crashOut);
+        try (BufferedReader reader = new BufferedReader(new FileReader(hs_err_file))) {
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ErrorHandling/TestDwarf.java:129]: Not sure this is the correct twr idiom when constructing multiple resources ie should it be:
```
try (var fr = new FileReader(hs_err_file);
      var reader = new BufferredReader(fr);) {
...
}
```

File: test/hotspot/jtreg/runtime/ErrorHandling/BadNativeStackInErrorHandlingTest.java [hunk 8]
```diff
@@ -58,44 +58,15 @@ public static void main(String[] args) throws Exception {
...
+    Pattern negativePatterns[] = {
+        Pattern.compile("\\[error occurred during error reporting \\(printing native stack\\), id .*\\]")
+    };
 
-    if (!lastLine.equals("END.")) {
-        throw new RuntimeException("hs-err file incomplete (missing END marker.)");
-    } else {
-        System.out.println("End marker found.");
-    }
+    HsErrFileUtils.checkHsErrFileContent(hs_err_file, null, negativePatterns, true, false);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/ErrorHandling/BadNativeStackInErrorHandlingTest.java:69]: Can we annotate the true/false parameters so the reader doesn't have to go and look up what they mean i.e. `true /* check END */, false /*not verbose */` ?

**Review by dholmes-ora — COMMENTED** [review]
Overall this seems quite reasonable. A few minor nits below.

Thanks.

File: test/hotspot/jtreg/runtime/ErrorHandling/BadNativeStackInErrorHandlingTest.java (hunk 8 above)
Comment (tstuefe) [reply test/hotspot/jtreg/runtime/ErrorHandling/BadNativeStackInErrorHandlingTest.java:69]: My IDE usually does this for me, but I can make it explicit, sure.

**Comment by tstuefe** [comment]
David, does this pass for you?

Any other review?

**Review by MBaesken — APPROVED** [review]
Looks good to me, minor points 
test/hotspot/jtreg/runtime/ErrorHandling/HsErrFileUtils.java

"existence of a set of pattern"  shouldn't this be
"existence of a set of patterns"  ?

please check COPYRIGHT headers , e.g.
test/hotspot/jtreg/runtime/ErrorHandling/NestedThreadsListHandleInErrorHandlingTest.java

**Comment by tstuefe** [comment]
Fixed comments as @MBaesken suggested; fixed copyrights. I took the freedom of adding SAP copyrights - with the correct introduction data - for all tests I personally introduces between 2013 and 2019. I had forgotten to do that.

**Comment by tstuefe** [comment]
@dholmes-ora are we good here?

**Review by dholmes-ora — APPROVED** [review]
Sorry for the delay.

One minor issue with a copyright line in one file that needs fixing, but otherwise this seems okay to me.

Thanks.

**Comment by dholmes-ora** [comment]
Running through our CI just to be cautious.

**Comment by dholmes-ora** [comment]
Our CI passed tiers 1-3

**Comment by tstuefe** [comment]
Thank you David!

I re-ran 32-bit tests on x86 and arm too.

/integrate

**Pull request closed without merging**